api/src/main/java/org/openmrs/api/impl/ConceptServiceImpl.java

```diff
@@ -1982,6 +1984,72 @@ public boolean hasAnyConceptAttribute(ConceptAttributeType conceptAttributeType)
 		return dao.getConceptAttributeCount(conceptAttributeType) > 0;
 	}
 
+	/**
+	 * @see org.openmrs.api.ConceptService#getConceptByNameOrIdOrMap(String conceptRef
+	 */
+	@Override
+	@Transactional(readOnly = true)
+	public Concept getConceptByNameOrIdOrMap(String conceptRef) throws APIException {
+		if (conceptRef == null) {
+			return null;
+		}
+		Concept cpt = null;
+		//check if input is a valid Uuid
+		if (!(conceptRef.length() < 36 || conceptRef.length() > 38 || conceptRef.contains(" ") || conceptRef.contains("."))) {
+			cpt = Context.getConceptService().getConceptByUuid(conceptRef);
+			if (cpt != null)
+				return cpt;
+		}
+		//check if input is map
+		if (conceptRef.indexOf(":") != -1) {
+			int idx = conceptRef.indexOf(":");
+			String conceptSource, conceptCode;
+			conceptCode = conceptRef.substring(idx + 1);
+			conceptSource = conceptRef.substring(0, idx);
+			cpt = Context.getConceptService().getConceptByMapping(conceptCode, conceptSource);
+			if (cpt != null)
+				return cpt;
+		}
+		//check if input is a name or id
+		if (!conceptRef.isEmpty()) {
+			//handle id
+			if (NumberUtils.isParsable(conceptRef)) {
+				int conceptId = Integer.valueOf(conceptRef);
+				cpt = Context.getConceptService().getConcept(conceptId);
+			} else
+				//handle name
+				cpt = Context.getConceptService().getConceptByName(conceptRef);
+			if (cpt != null)
+				return cpt;
+		}
+		if (conceptRef.contains(".")) {
+			return getConceptByNameOrIdOrMap(evaluateStaticConstant(conceptRef));
+		}
```

Comment:
Why are we adding this functionality? This doesn't actually seem very useful. What is the use-case where I know that the concept I want is stored in `org.openmrs.my.package.MyGreatClass.MY_GREAT_CONCEPT` but I can't just call `Context.getConceptService().getConceptByConceptReference(MyGreatClass.MY_GREAT_CONCEPT)`?

Reply:
Hello @ibacher!   If don't use the method `evaluateStaticConstant(String)` and try to get a concept by static constant (with reference to `org.openmrs.my.package.MyGreatClass.MY_CLEAR_CONCEPT`) like;

> if (conceptRef.contains(".")) {
			return getConceptByReference(MyNiceClass.MY_CLEAR_CONCEPT);
     }
> 
i get a `java.lang.NoClassDefFoundError: org/apache/logging/log4j/core/Appender` Error.  Actually, the method `evaluateStaticConstant(String)` loads the class path and retrieves the required concept.

Reply:
@jnsereko Ok, but I'm unclear on why were trying to load something given the string "org.openmrs.my.package.MyGreatClass.MY_GREAT_CONCEPT" at all. Is there some use case for this?

Reply:
Hello @ibacher! thank you for your support.  The truth is that i haven't seen it being used anywhere, except that the method,[ HtmlFormEntryUtil#getConcept (id)](https://github.com/openmrs/openmrs-module-htmlformentry/blob/master/api/src/main/java/org/openmrs/module/htmlformentry/HtmlFormEntryUtil.java#L564-L605) (which the description on the issue pointed to) had this functionality.  As a result, i thought that we could need it here.  
cc @samuelmale @bmamlin @dkayiwa

Reply:
@jnsereko Thanks! That makes things very clear!